docs/proposal-platforms.md

```diff
@@ -0,0 +1,259 @@
+    file (i.e. bottom up). This is opposite to the method currently used for
+    matching hosts but is required because any additional platforms defined in
+    user config files get added to the end of the list. Currently it is
+    impossible to add hosts defined as regular expressions in the user config if
+    there is already a more generic regular expression defined in the global
+    config.
+  * There will be no special treatment of sections which define a single
+    platform. The first matching section found will be used.
+  * Note that the first reference to a section in the global or user config
+    determines its position in the list. You can repeat a section later in the
+    config and alter some settings but this doesn't change its position.
+
+* The names of all new and existing settings are subject to change until we
+  release cylc 8.0. Initially names will be chosen which match current usage
+  wherever possible. We intend to review all the config settings once we have
+  the key functionality in place.
+
+* The default platform is `localhost`. At cylc 7, any settings for `localhost`
+  act as defaults for all other hosts but we do **not** propose doing this for
+  platforms. The downside is more duplication if you need to override a
+  particular setting for every platform but we think this is clearer and safer.
+
+* If `hosts` is defined then the platform section heading should specify a single
+  platform name - otherwise you are defining lots of identical platforms which
+  makes no sense (although it is relatively harmless so we probably won't try to
+  prevent this or give any warnings if this happens).
+  * As a future enhancement we could support the use of string substitution using
+    [re.sub](https://docs.python.org/3/library/re.html#re.sub) (this would make
+    the deprecation logic more complicated so we don't propose doing this at
+    first). e.g. something like:
+```ini
+[platforms]
+    [[hpcl[12]-bg]]
+        hosts = ("-bg$","")
+        retrieve job logs = True
+        batch system = background
+```
+
+## Example platform usage
+
+These are examples of how the platform may be assigned in the `[runtime]`
+section for a job. They refer to the example platforms shown above.
+
+`platform = desktop01`
+* The simplest example. The job will be run on the host `desktop01`.
+
+`platform = special`
+* There is no platform named `special` defined so this will fail validation.
+
+`platform = hpc`
+* The host used will be chosen at random.
+
+`platform = hpcl1-bg, hpcl2-bg`
+* The platform used will be chosen at random.
+
+`platform = hpc-bg`
+* This is equivalent to the previous example.
+
+`platform = $(rose host-select linux)`
+* No checking can be performed at validation so if the command does not return a
+  valid platform this will result in a submit failure.
+
+## Deprecation method
+
+* You can't mix old with new. If `platform` is defined for **any** task then any
```

Comment:
So this would mean we can't override settings at the task level e.g:

```ini

[foo]
    platform = bar
    [[platform]]
        host = bar01  # hardcode a login node
```

This is probably a good thing the user should just go away and create a new platform with the modification.

Under this proposal, in particular the system of effectively naming platforms after their regexes, how could I duplicate a platform to apply a small diff?